### PR TITLE
Implement head_mask for Flax BERT and other models copied from BERT

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?

This PR implements attention head masking for Flax BERT models, and other models copied from BERT, i.e. Roberta, BigBird, Electra. This PR thus narrows the gap between Flax and other implementations.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@LysandreJik

### CODE REVIEW

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 1]
```diff
@@ -234,7 +240,14 @@ def setup(self):
             kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
         )
 
-    def __call__(self, hidden_states, attention_mask, deterministic=True, output_attentions: bool = False):
+    def __call__(
+        self,
+        hidden_states,
+        attention_mask,
+        layer_head_mask=None,
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:247]: It should be `None` here since we always initialize the mask in the `PreTrained` module.

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 2]
```diff
@@ -310,12 +327,23 @@ def setup(self):
         self.self = FlaxBertSelfAttention(self.config, dtype=self.dtype)
         self.output = FlaxBertSelfOutput(self.config, dtype=self.dtype)
 
-    def __call__(self, hidden_states, attention_mask, deterministic=True, output_attentions: bool = False):
+    def __call__(
+        self,
+        hidden_states,
+        attention_mask,
+        layer_head_mask=None,
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:334]: same comment as above

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 3]
```diff
@@ -310,12 +327,23 @@ def setup(self):
...
+        output_attentions: bool = False,
+    ):
         # Attention mask comes in as attention_mask.shape == (*batch_sizes, kv_length)
         # FLAX expects: attention_mask.shape == (*batch_sizes, 1, 1, kv_length) such that it is broadcastable
         # with attn_weights.shape == (*batch_sizes, num_heads, q_length, kv_length)
         attn_outputs = self.self(
-            hidden_states, attention_mask, deterministic=deterministic, output_attentions=output_attentions
+            hidden_states,
+            attention_mask,
+            layer_head_mask=layer_head_mask,
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:344]: same comment as above

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 4]
```diff
@@ -615,6 +670,9 @@ def __call__(
         if attention_mask is None:
             attention_mask = jnp.ones_like(input_ids)
 
+        if head_mask is None:
+            head_mask = jnp.ones([self.config.num_hidden_layers, self.config.num_attention_heads])
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:674]:
```suggestion
            head_mask = jnp.ones((self.config.num_hidden_layers, self.config.num_attention_heads))
```

File: src/transformers/models/big_bird/modeling_flax_big_bird.py [hunk 5]
```diff
@@ -1113,14 +1130,32 @@ def setup(self):
...
+                deterministic=deterministic,
+                output_attentions=output_attentions,
+            )
+        else:
+            attn_outputs = self.self(
+                hidden_states,
+                attention_mask,
+                deterministic=deterministic,
+                output_attentions=output_attentions,
+            )
```
Comment (patil-suraj) [inline src/transformers/models/big_bird/modeling_flax_big_bird.py:1158]: why don't we use the head_mask here?

File: src/transformers/models/big_bird/modeling_flax_big_bird.py [hunk 6]
```diff
@@ -1423,6 +1485,7 @@ def __call__(
             jnp.array(attention_mask, dtype="i4"),
             jnp.array(token_type_ids, dtype="i4"),
             jnp.array(position_ids, dtype="i4"),
+            head_mask,
```
Comment (patil-suraj) [inline src/transformers/models/big_bird/modeling_flax_big_bird.py:1488]: we should initialize the `head_mask` if it's `None` here. Is there any reason not to do that here?

File: src/transformers/models/electra/modeling_flax_electra.py [hunk 7]
```diff
@@ -552,6 +607,7 @@ def __call__(
             jnp.array(attention_mask, dtype="i4"),
             jnp.array(token_type_ids, dtype="i4"),
             jnp.array(position_ids, dtype="i4"),
+            head_mask,
```
Comment (patil-suraj) [inline src/transformers/models/electra/modeling_flax_electra.py:610]: same comment as above. We should initialize the `head_mask` if it's `None`.

File: src/transformers/models/roberta/modeling_flax_roberta.py [hunk 8]
```diff
@@ -595,6 +650,7 @@ def __call__(
             jnp.array(attention_mask, dtype="i4"),
             jnp.array(token_type_ids, dtype="i4"),
             jnp.array(position_ids, dtype="i4"),
+            head_mask,
```
Comment (patil-suraj) [inline src/transformers/models/roberta/modeling_flax_roberta.py:653]: same comment as above. We should initilize the `head_mask` if it's `None`

File: tests/test_modeling_flax_common.py [hunk 9]
```diff
@@ -777,6 +778,53 @@ def test_save_load_in_bf16(self):
...
+    def test_headmasking(self):
+        if not self.test_head_masking:
+            return
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+        config.return_dict = True
+
+        def _prepare_layer_head_mask(i, attention_heads, num_hidden_layers):
+            if i == 0:
+                return np.concatenate([np.zeros(1, dtype=jnp.int32), np.ones(attention_heads - 1, dtype=jnp.int32)])
+            if i == num_hidden_layers - 1:
```
Comment (patil-suraj) [inline tests/test_modeling_flax_common.py:790]: Nice!

File: tests/test_modeling_flax_roberta.py [hunk 10]
```diff
@@ -116,6 +116,8 @@ def prepare_config_and_inputs_for_common(self):
 @require_flax
 class FlaxRobertaModelTest(FlaxModelTesterMixin, unittest.TestCase):
 
+    test_head_masking = True
+
```
Comment (patil-suraj) [inline tests/test_modeling_flax_roberta.py:120]: We should also test BigBird, no ?

**Review by patil-suraj — COMMENTED** [review]
Sorry to only reply now. It looks good to me! I've left a few comments. Specifically, 

- the `head_mask` should not be `None` in the submodules, as we always initialize it in the `PreTrainedModel`.
- BigBird is not tested for `head_mask`, is not working?

Thanks a lot for adding this feature @stancld!

File: src/transformers/models/big_bird/modeling_flax_big_bird.py (hunk 5 above)
Comment (stancld) [reply src/transformers/models/big_bird/modeling_flax_big_bird.py:1158]: Actually, head masking is not used for this type of attention in `torch` implementation either. However, I think this should be somehow handled as well. @vasudevgupta7 - I know you contributed `BigBird` to `transformers`. Do you have any idea, how this should be handled?

File: tests/test_modeling_flax_roberta.py (hunk 10 above)
Comment (stancld) [reply tests/test_modeling_flax_roberta.py:120]: BigBird is not testing for head masking for `torch` either, as it's required different test script because of various attention mechanisms. Maybe, I can have a look at this in another PR?

**Comment by stancld** [comment]
Hi @patil-suraj, thank you a lot for the review. I commented on the issue with BigBird head-masking. As I mentioned, I tried to follow the current `PyTorch` implementation, and I'm definitely willing to have a look at handling head masking for different attention mechanisms.

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 11]
```diff
@@ -718,6 +779,7 @@ def __call__(
         attention_mask,
         token_type_ids,
         position_ids,
+        head_mask: Optional[np.ndarray] = None,
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:782]:
```suggestion
        head_mask,
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 12]
```diff
@@ -810,6 +873,7 @@ def __call__(
...
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:876]:
```suggestion
        head_mask,
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 13]
```diff
@@ -870,6 +935,7 @@ def __call__(
...
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:938]:
```suggestion
        head_mask,
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 14]
```diff
@@ -883,6 +949,7 @@ def __call__(
             attention_mask,
             token_type_ids,
             position_ids,
+            head_mask=head_mask,
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:952]:
```suggestion
            head_mask,
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 15]
```diff
@@ -962,6 +1029,7 @@ def __call__(
         attention_mask,
         token_type_ids,
         position_ids,
+        head_mask: Optional[np.ndarray] = None,
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:1032]:
```suggestion
        head_mask,
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 16]
```diff
@@ -1028,6 +1097,7 @@ def __call__(
...
```
Comment (patil-suraj) [inline src/transformers/models/bert/modeling_flax_bert.py:1100]:
```suggestion
        head_mask,
```

**Review by patil-suraj — APPROVED** [review]
Looks good now! @stancld Would be awesome if you could make every `head_mask` argument required except in the `PreTrained` class, then we should be good for merge!

File: tests/test_modeling_flax_roberta.py (hunk 10 above)
Comment (patil-suraj) [reply tests/test_modeling_flax_roberta.py:120]: I see, thanks!

**Comment by stancld** [comment]
@patil-suraj Thanks a lot for the review and suggestions, it should be fixed now :]

**Comment by patil-suraj** [comment]
I forgot to write this earlier, do you think you could also update the flax templates, the templates test are failing, 
If not let me know, I will take care of it , thanks :)

**Comment by stancld** [comment]
@patil-suraj Oh, good point. I'll update the template :]
